DEP: Deprecate `distance.wminkowski`

peterbell10 opened the pull request:
#### Reference issue
See gh-7893

#### What does this implement/fix?
`wminkowski` was "soft deprecated" in gh-7905 (SciPy 1.0) and, as per gh-7893, should have started raising a warning in SciPy 1.3. This (belatedly) marks it as fully deprecated and also repurposes the native `cdist` and `pdist` implementation for `wminkowski` so that it works for weighted `minkowski`. That way there is no performance regression for users switching over to `minkowski` from `wminkowski`.

rgommers commented:
> also repurposes the native `cdist` and `pdist` implementation for `wminkowski` so that it works for weighted `minkowski`. That way there is no performance regression for users switching over to `minkowski` from `wminkowski`.

Sounds good, but I'm not sure I fully understand this. You are adding C code in this PR with the comment that it must be removed again in 1.8.0 - how can this be needed? Either users switch to using `minkowski`, which already supports weights (also in `pdist`, `cdist`) so doesn't need changing _temporarily_ (new C code for performance reasons must be compatible with current `minkowski` behavior), or users stay with `kminkowski` and no new code is needed either.

peterbell10 commented:
>You are adding C code in this PR with the comment that it must be removed again in 1.8.0 - how can this be needed?

The diff is perhaps a bit confusing here. The current `{p,c}dist_weighted_minkowski` corresponds to the `wminkowski` metric, with it's definition of the weights. I've renamed it to `{p,c}dist_old_weighted_minkowski`. That's the existing code that should be removed in SciPy 1.8.0.

I've also added a weighted variant of `minkowski` to the C code, which is taking over the names `{p,c}dist_weighted_minkowski`. That is the new code that won't be removed in 1.8. However, because I've given it the same name as the old function and it's mostly boilerplate anyway, the new function doesn't really appear in the diff.

>users switch to using minkowski, which already supports weights

Yes, but if you specify weights it doesn't go through the C code at the moment. Instead it falls back to the slow python implementation. I generally think you shouldn't deprecate functionality if the alternative is not as good. So I'm making `minkowski` with weights as fast as `wminkowski`.

rgommers commented:
> However, because I've given it the same name as the old function and it's mostly boilerplate anyway, the new function doesn't really appear in the diff.

Ah yes, that explains it, thanks. I'll review in more detail.




> > You are adding C code in this PR with the comment that it must be removed again in 1.8.0 - how can this be needed?
> 
> The diff is perhaps a bit confusing here. The current `{p,c}dist_weighted_minkowski` corresponds to the `wminkowski` metric, with it's definition of the weights. I've renamed it to `{p,c}dist_old_weighted_minkowski`. That's the existing code that should be removed in SciPy 1.8.0.
> 
> I've also added a weighted variant of `minkowski` to the C code, which is taking over the names `{p,c}dist_weighted_minkowski`. That is the new code that won't be removed in 1.8. However, because I've given it the same name as the old function and it's mostly boilerplate anyway, the new function doesn't really appear in the diff.
> 
> > users switch to using minkowski, which already supports weights
> 
> Yes, but if you specify weights it doesn't go through the C code at the moment. Instead it falls back to the slow python implementation. I generally think you shouldn't deprecate functionality if the alternative is not as good. So I'm making `minkowski` with weights as fast as `wminkowski`.

Yep, completely agree.

rgommers commented:
@peterbell10 changes look good, but this PR has a bad interaction with gh-12375 (add fast path for `p = 1, 2` for `minkowski`). If I rebase this PR on master, I see the following failures:
```
FAILED tests/test_distance.py::TestCdist::test_cdist_minkowski_random[inf]
FAILED tests/test_distance.py::TestPdist::test_pdist_minkowski_random_p[inf]
```
Could you rebase and update?

peterbell10 commented:
This should be fixed now. I had to add a special case for `p=inf` which dispatches to a new native weighted Chebyshev distance implementation.

rgommers reviewed (approved):
This seems all happy now, in it goes. Thanks Peter!

rgommers commented:
I added a release note for this.
